### PR TITLE
Refactor code for service seeking code challenge

### PR DESCRIPTION
**Pull request by jeffreycastro**
- [x] concept for Todo
- [x] concept for TodoList
- [x] concept for User

### CODE REVIEW

**Comment by kennethcalamay** [comment]
Hey Jeffrey, checking your submission.

File: app/concepts/todo/create.rb [hunk 1]
```diff
@@ -1,50 +1,30 @@
...
-  #     else
-  #         create the `todo`
-  #         create a `user`
-  #         create a default `todo_list`
-  #         associate `todo_list` with `user`
-  #         associate `todo` with `todo_list`
-  def process(params)
-    # make sure title is not blank
-    return invalid! if params[:todo][:title].blank?
+  contract do
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:5]: Good job for cleaning up the comments.

File: app/concepts/todo/create.rb [hunk 2]
```diff
@@ -1,50 +1,30 @@
...
+    property :title
+    property :todo_list_id
+
+    validates :title, presence: true
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:10]: Good to see use of contracts.

File: app/concepts/todo/create.rb [hunk 3]
```diff
@@ -1,50 +1,30 @@
...
 
-    # build the todo
-    todo = @model = Todo.new(params.require(:todo).permit(:title, :description))
+  def process(params)
+    validate(params[:todo]) do |f|
+      f.todo_list_id = associate_list(params[:todo_list_id])
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:14]: Is associate_list associating the list or fetching the list?

File: app/concepts/todo/create.rb [hunk 4]
```diff
@@ -1,50 +1,30 @@
...
-      todo_list = TodoList.find(params[:todo_list_id])
-    end
+    default_todo = TodoList.where(name: 'Default To-do List').first
+    return default_todo.id if default_todo.present?
 
-    # associate the todo with the todo_list
-    todo.list = todo_list
-    todo.save!
+    todo_list_attr = { name: 'Default To-do List' }
+    TodoList::Create.(todo_list: todo_list_attr).model.id
```
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:28]: Should all these logic belong to this class?

File: app/concepts/todo/form_cell.rb [hunk 5]
```diff
@@ -21,11 +20,24 @@ def default_prompt
...
+    TodoList.find(params[:todo_list_id]) rescue nil
   end
 
   def action_url
+    return parent_controller.todo_list_todos_path if todo_list.present?
     parent_controller.todos_path
   end
 
+  def back_url
+    return parent_controller.todo_list_path(todo_list) if todo_list.present?
```
Comment (kennethcalamay) [inline app/concepts/todo/form_cell.rb:32]: Nice!

File: app/concepts/todo_list/create.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  contract do
+    property :name
+    property :user_id
+
+    validates :name, presence: true
+  end
+
+  def process(params)
+    validate(params[:todo_list]) do |f|
+      f.user_id = associate_user(params)
```
Comment (kennethcalamay) [inline app/concepts/todo_list/create.rb:14]: Is associate_user associating a todo_list with a user? Or is it just fetching?

File: app/concepts/todo_list/create.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  def associate_user(params)
+    return params[:current_user_id] if params[:current_user_id].present?
+    if params[:current_user].nil?
+      user_attr = { fullname: 'Guest' }
+      user = User::Create.(user: user_attr).model.id
+    else
+      user = params[:current_user].id
+    end
+  end
```
Comment (kennethcalamay) [inline app/concepts/todo_list/create.rb:29]: Do all this logic belong here?

File: app/concepts/todo_list/form_cell.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+class TodoList::FormCell < Cell::Concept
```
Comment (kennethcalamay) [inline app/concepts/todo_list/form_cell.rb:1]: 👍

File: app/concepts/user/create.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+    validates :fullname, presence: true
+  end
+
+  def process(params)
+    validate(params[:user]) do |f|
+      f.save
+    end
+  end
+end
```
Comment (kennethcalamay) [inline app/concepts/user/create.rb:16]: 👍

File: app/controllers/todo_lists_controller.rb [hunk 10]
```diff
@@ -8,10 +7,21 @@ def show
...
+  end
+
+  def create
+    res = run TodoList::Create do
+      return redirect_to @model, notice: "List was successfully created!"
+    end
+
+    @errors = res.errors.full_messages
+    render :new
+  end
```
Comment (kennethcalamay) [inline app/controllers/todo_lists_controller.rb:20]: 👍

File: spec/concepts/todo/create_spec.rb [hunk 11]
```diff
@@ -1,7 +1,46 @@
...
+  context "invalid" do
+    it "does not create a todo" do
+      attrs = { title: nil }
+      res, op = Todo::Create.run(todo: attrs)
+      expect(res).to eq(false)
+      expect(op.model).not_to be_persisted
+
+      msg = op.contract.errors.full_messages
+      expect(msg).to include("Title can't be blank")
+    end
```
Comment (kennethcalamay) [inline spec/concepts/todo/create_spec.rb:44]: Good command of RSpec.

File: spec/concepts/todo_list/create_spec.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      attrs = { name: nil }
+      res, op = TodoList::Create.run(todo_list: attrs)
+      expect(res).to eq(false)
+      expect(op.model).not_to be_persisted
+
+      msg = op.contract.errors.full_messages
+      expect(msg).to include("Name can't be blank")
+    end
+  end
+end
```
Comment (kennethcalamay) [inline spec/concepts/todo_list/create_spec.rb:56]: 👍

File: spec/concepts/user/create_spec.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    it "does not create a user" do
+      attrs = { fullname: nil }
+      res, op = User::Create.run(user: attrs)
+      expect(res).to eq(false)
+      expect(op.model).not_to be_persisted
+
+      msg = op.contract.errors.full_messages
+      expect(msg).to include("Fullname can't be blank")
+    end
+  end
```
Comment (kennethcalamay) [inline spec/concepts/user/create_spec.rb:24]: 👍

File: app/concepts/todo/create.rb (hunk 3 above)
Comment (jeffreycastro) [inline app/concepts/todo/create.rb:14]: Its actually just fetching the list id, sorry for the misleading method name.  
Though in the `associate_list` method, I am doing some logic that is not just fetching.

File: app/concepts/todo/create.rb (hunk 4 above)
Comment (jeffreycastro) [inline app/concepts/todo/create.rb:28]: I just didnt want to put an if-else logic inside the `process` method.  
Can you suggest a better location for this logic?

File: app/concepts/todo_list/create.rb (hunk 6 above)
Comment (jeffreycastro) [inline app/concepts/todo_list/create.rb:14]: This is similar to the todo list, it is just fetching the id of the user

File: app/concepts/todo/create.rb (hunk 4 above)
Comment (kennethcalamay) [inline app/concepts/todo/create.rb:28]: Yeah. Creation of the todolist and whatever logic it needs to do it should be inside the TodoList::Create class.

**Pull request closed without merging**